src/entt/entity/polymorphic.hpp

```diff
@@ -0,0 +1,554 @@
+};
+
+/**
+ * @brief Holds pointer to child type storage and a function to convert it into parent type
+ * @tparam Entity underlying entity type
+ * @tparam Type polymorphic component type to be converted into
+ */
+template<typename Entity, typename Type>
+class poly_pool_holder : public poly_pool_holder_base<Entity> {
+public:
+    using poly_pool_holder_base<Entity>::poly_pool_holder_base;
+
+    /** @brief entity type of the component pools */
+    using entity_type = Entity;
+    /** @brief component type */
+    using value_type = Type;
+    /** @brief pointer to the component, will be same as value_type, if component itself is a pointer */
+    using pointer_type = std::remove_pointer_t<Type>*;
+    /** @copydoc pointer_type */
+    using const_pointer_type = const std::remove_pointer_t<Type>*;
+
+    /**
+     * @brief gets child from a child pool as parent type reference by given entity
+     * @param ptr pointer from the child type set
+     * @return Returns reference to Type, converted from a given pointer. Will return pointer instead of reference, if parent type is a pointer.
+     */
+    inline pointer_type try_get(const Entity ent) ENTT_NOEXCEPT {
+        return pointer_type(this->getter_ptr(this->pool_ptr, ent));
+    }
+
+    /** @copydoc try_get */
+    inline const_pointer_type try_get(const Entity ent) const ENTT_NOEXCEPT {
+        return const_cast<poly_pool_holder<Entity, Type>*>(this)->try_get(ent);
+    }
+
+    /**
+     * @brief removes component from pool by entity
+     * @param ent entity
+     * @return true, if component was removed, false, if it didnt exist
+     */
+    inline bool remove(const Entity ent) {
+        return this->remover_ptr(this->pool_ptr, ent);
+    }
+
+    /** @brief returns underlying pool */
+    inline auto& pool() ENTT_NOEXCEPT {
+        return *this->pool_ptr;
+    }
+
+    /** @copydoc pool */
+    inline const auto& pool() const ENTT_NOEXCEPT {
+        return *this->pool_ptr;
+    }
+};
+
+/**
+ * @brief Holds runtime information about one polymorphic component type.
+ * @tparam Entity
+ */
+template<typename Entity, typename Type>
+class poly_type {
+    /** @brief internally used to convert base_poly_pool_holder to a pool_poly_holder<Entity, Type> during the iteration */
+    struct cast_pool_holder {
+        auto& operator()(poly_pool_holder_base<Entity>& holder) {
+            return static_cast<poly_pool_holder<Entity, Type>&>(holder);
+        }
+
+        const auto& operator()(const poly_pool_holder_base<Entity>& holder) const {
+            return static_cast<const poly_pool_holder<Entity, Type>&>(holder);
+        }
+    };
+
+    /** @brief derived value to base pointer conversion, workaround for if constexpr bug on some compilers */
+    template<typename Base, typename Derived, typename = void>
+    struct derived_to_base_ptr {
+        inline static Base* convert(Derived& ref) ENTT_NOEXCEPT {
+            return static_cast<Base*>(std::addressof(ref));
```

Comment:
This is the workaround for `if constexpr` bug on MSVC v141, mentioned in the comment above